File: crunch/src/test/java/org/apache/crunch/lib/SecondarySortTest.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.crunch.lib;

import static org.apache.crunch.types.avro.Avros.*;
import static org.junit.Assert.assertEquals;

import org.apache.crunch.DoFn;
import org.apache.crunch.Emitter;
import org.apache.crunch.PCollection;
import org.apache.crunch.PTable;
import org.apache.crunch.Pair;
import org.apache.crunch.impl.mem.MemPipeline;
import org.junit.Test;

import com.google.common.collect.ImmutableList;


public class SecondarySortTest {
  @Test
  public void testInMemory() throws Exception {
    PTable<Long, Pair<Long, String>> input = MemPipeline.typedTableOf(tableOf(longs(), pairs(longs(), strings())),
        1729L, Pair.of(17L, "a"), 100L, Pair.of(29L, "b"), 1729L, Pair.of(29L, "c"));
    PCollection<String> letters = SecondarySort.sortAndApply(input, new StringifyFn(), strings());
    assertEquals(ImmutableList.of("b", "ac"), letters.materialize());
  }
  
  private static class StringifyFn extends DoFn<Pair<Long, Iterable<Pair<Long, String>>>, String> {
    @Override
    public void process(Pair<Long, Iterable<Pair<Long, String>>> input, Emitter<String> emitter) {
      StringBuilder sb = new StringBuilder();
      for (Pair<Long, String> p : input.second()) {
        sb.append(p.second());
      }
      emitter.emit(sb.toString());
    }
  }
}

File: crunch-scrunch/src/it/scala/org/apache/crunch/scrunch/SecondarySortTest.scala
/*
 * *
 *  * Licensed to the Apache Software Foundation (ASF) under one
 *  * or more contributor license agreements.  See the NOTICE file
 *  * distributed with this work for additional information
 *  * regarding copyright ownership.  The ASF licenses this file
 *  * to you under the Apache License, Version 2.0 (the
 *  * "License"); you may not use this file except in compliance
 *  * with the License.  You may obtain a copy of the License at
 *  *
 *  *     http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *
 */

package org.apache.crunch.scrunch

import org.junit.Assert._
import org.junit.Test

class SecondarySortTest extends CrunchSuite with Serializable {

  @Test def testSecondarySortAvros {
    runSecondarySort(Avros)
  }

  @Test def testSecondarySortWritables {
    runSecondarySort(Writables)
  }

  def runSecondarySort(ptf: PTypeFamily) {
    val p = Pipeline.mapReduce(classOf[SecondarySortTest], tempDir.getDefaultConfiguration)
    val inputFile = tempDir.copyResourceFileName("secondary_sort_input.txt")
    val lines = p.read(From.textFile(inputFile, ptf.strings)).map(input => {
      val pieces = input.split(",")
      (pieces(0), (pieces(1).trim.toInt, pieces(2).trim.toInt))
    })
    .secondarySortAndMap((k, iter: Iterable[(Int, Int)]) => {
      List(k, iter.mkString(",")).mkString(",")
    })
    .materialize
    assertEquals(List("one,(-5,10),(1,1),(2,-3)", "three,(0,-1)", "two,(1,7),(2,6),(4,5)"), lines.toList)
    p.done
  }
}
